Split declarations from definitions

chelini opened the pull request.

wsmoses reviewed (approved):
A very needed cleanup

wsmoses commented:
Something in this PR broke the tests. To ease discovery of what happened, I reverted this PR and added the test commit first in #16 so tests can be verified before this PR is applied.

chelini commented:
> Something in this PR broke the tests. To ease discovery of what happened, I reverted this PR and added the test commit first in #16 so tests can be verified before this PR is applied.

Sorry I will look into it tomorrow. In any case, on my laptop all the test breaks with the following error:

```
Command Output (stderr):
--
warning: argument unused during compilation: '--cuda-gpu-arch=sm_35'
/Users/lorenzo/clang-mlir/mlir/tools/mlir-clang/Test/polybench/medley/nussinov/nussinov.c:13:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
Assertion failed: (Index < Length && "Invalid index!"), function operator[], file /Users/lorenzo/clang-mlir/llvm/include/llvm/ADT/ArrayRef.h, line 248.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/lorenzo/clang-mlir/build/bin/mlir-clang /Users/lorenzo/clang-mlir/mlir/tools/mlir-clang/Test/polybench/medley/nussinov/nussinov.c main -I /Users/lorenzo/clang-mlir/llvm/../clang/lib/Headers -I /Users/lorenzo/clang-mlir/mlir/tools/mlir-clang/Test/polybench/utilities 
1.	<eof> parser at end of file
0  mlir-clang               0x000000010c4649c8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 40
1  mlir-clang               0x000000010c463798 llvm::sys::RunSignalHandlers() + 248
2  mlir-clang               0x000000010c464fbc SignalHandler(int) + 268
3  libsystem_platform.dylib 0x00007fff6c6e65fd _sigtramp + 29
4  mlir-clang               0x000000010ea4bbac llvm::InstrProfError::ID + 108460
5  libsystem_c.dylib        0x00007fff6c5bc808 abort + 120
6  libsystem_c.dylib        0x00007fff6c5bbac6 err + 0
7  mlir-clang               0x000000010dd158a3 MLIRScanner::VisitCallExpr(clang::CallExpr*) (.cold.17) + 35
8  mlir-clang               0x000000010c237170 MLIRScanner::VisitCallExpr(clang::CallExpr*) + 6064
9  mlir-clang               0x000000010c230c39 clang::StmtVisitorBase<std::__1::add_pointer, MLIRScanner, ValueWithOffsets>::Visit(clang::Stmt*) + 249
10 mlir-clang               0x000000010c23302b MLIRScanner::VisitCompoundStmt(clang::CompoundStmt*) + 171
11 mlir-clang               0x000000010c230e49 clang::StmtVisitorBase<std::__1::add_pointer, MLIRScanner, ValueWithOffsets>::Visit(clang::Stmt*) + 777
12 mlir-clang               0x000000010c2305d9 MLIRScanner::MLIRScanner(MLIRASTConsumer&, mlir::FuncOp, clang::FunctionDecl const*, mlir::ModuleOp&) + 1497
13 mlir-clang               0x000000010c22ff75 MLIRASTConsumer::run() + 421
14 mlir-clang               0x000000010c22f168 MLIRASTConsumer::HandleTopLevelDecl(clang::DeclGroupRef) + 456
15 mlir-clang               0x000000010cde0212 clang::ParseAST(clang::Sema&, bool, bool) + 514
16 mlir-clang               0x000000010cb67e93 clang::FrontendAction::Execute() + 163
17 mlir-clang               0x000000010c22af66 main + 5398
18 libdyld.dylib            0x00007fff6c4edcc9 start + 1
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /Users/lorenzo/clang-mlir/build/bin/FileCheck /Users/lorenzo/clang-mlir/mlir/tools/mlir-clang/Test/polybench/medley/nussinov/nussinov.c

--
```

I am running on a MacOs Catalina. Do you know any easy fix?

wsmoses commented:
My guess is that one of the headers or other definition is different on macos and triggering an error. For ease rather than having the test try to generate mlir for main (and its dependencies) what happens when you try on the actual kernel.

wsmoses commented:
Also possibly not related to the assertion the include error happens because the default clang install on MacOS requires special build flags to get certain headers working. Can you try -I'ing the directory on your laptop that contains stdio.h?

chelini commented:
> Also possibly not related to the assertion the include error happens because the default clang install on MacOS requires special build flags to get certain headers working. Can you try -I'ing the directory on your laptop that contains stdio.h?

Thanks. I managed to make it work. I will check why the commits failed tomorrow.
